src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,347 @@
+      auto itb = rand_map.emplace(u, 0);
+      if (itb.second) {
+        auto ng = ng0;
+        ng.discard(u);
+        if (num_seeds > 1) {
+          norm.reset();
+          auto rnd = a0 * norm(ng);
+          ng = ng1;
+          ng.discard(u);
+          norm.reset();
+          rnd += a1 * norm(ng);
+          itb.first->second = std::erfc(-rnd * (FloatType)/*M_SQRT1_2*/0.70710678118654752440) / 2;
+        } else {
+          uni.reset();
+          itb.first->second = uni(ng);
+        }
+      }
+      const auto rnd = itb.first->second;
+      num_edges +=
+          rnd <= (importance_sampling - weights ? c * hop_map[v] : c * (weights ? A[j] : 1));
+      rands[off++] = rnd;
+    }
+  }
+
+  IdArray picked_row = NDArray::Empty({num_edges},
+                                      DGLDataType{kDGLInt, 8*sizeof(IdxType), 1},
+                                      ctx);
+  IdArray picked_col = NDArray::Empty({num_edges},
+                                      DGLDataType{kDGLInt, 8*sizeof(IdxType), 1},
+                                      ctx);
+  IdArray picked_idx = NDArray::Empty({num_edges},
+                                      DGLDataType{kDGLInt, 8*sizeof(IdxType), 1},
+                                      ctx);
+  FloatArray importances = NDArray::Empty({importance_sampling ? num_edges : 1},
+                                          DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                          ctx);
+  IdxType* picked_rdata = static_cast<IdxType*>(picked_row->data);
+  IdxType* picked_cdata = static_cast<IdxType*>(picked_col->data);
+  IdxType* picked_idata = static_cast<IdxType*>(picked_idx->data);
+  FloatType* importances_data = static_cast<FloatType*>(importances->data);
+
+  std::size_t off = 0, idx = 0;
+
+  for (int64_t i = 0; i < num_rows; ++i) {
+    const IdxType rid = rows_data[i];
+    const auto d = indptr[rid + 1] - indptr[rid];
+    if (d == 0)
+      continue;
+    const auto c = cs[i];
+
+    FloatType norm_inv_p = 0;
+    const auto off_start = off;
+    for (auto j = indptr[rid]; j < indptr[rid + 1]; j++) {
+      const auto v = indices[j];
+      const auto w = (weights ? A[j] : 1);
+      const auto ps = std::min(ONE, importance_sampling - weights ? c * hop_map[v] : c * w);
+      const auto rnd = rands[idx++];
+      if (rnd <= ps) {
+        norm_inv_p += w / ps;
+        picked_rdata[off] = rid;
+        picked_cdata[off] = indices[j];
+        picked_idata[off] = data ? data[j] : j;
+        if (importance_sampling)
+          importances_data[off] = w / ps;
+        off++;
+      }
+    }
+
+    if (importance_sampling) {
+      for (auto i = off_start; i < off; i++)
+        // so that fn.mean can be used
+        importances_data[i] *= (off - off_start) / norm_inv_p;
+    }
+  }
+
+  assert((IdxType)off == num_edges);
+
+  return std::make_pair(COOMatrix(mat.num_rows, mat.num_cols,
+                   picked_row, picked_col, picked_idx), importances);
+}
```

Comment:
Could we break up this function into smaller functions?  For instance, I feel the `if (importance_sampling)` block can be put into a separate function, and L211-258 can be put into another, L279-309 feels like another.

A rule of thumb is to have each function do one thing and do it well.  If a function is doing many things at once, break it into smaller functions.